File: 4newtonInterpolation.r
X11()

inter <- function(file){
    #input:
    df1 <- read.table(file, col.names=c("x", "y"))

    #finite divided differences:
    n <- length(df1$x)
    M <- matrix(0, nrow=n, ncol=n)
    M[,1] <- df1$y
    for(j in 1:(n-1))
        for(i in 0:(n-j-1))
            M[i+1,j+1] = (M[i+2,j] - M[i+1,j]) /
                (df1$x[i+j+1] - df1$x[i+1]);

   #interpolation:
   x <- seq(df1$x[1], df1$x[n], by=0.01)
   y <- M[1,1]
   factor <- 1
   for(j in 1:(n-1)){
       factor <- factor * (x-df1$x[j])
       y <- y + factor * M[1,j+1]
   }

   #output:
   write.table(data.frame(x, y),
    file="out2.txt", sep='\t', row.names=F, col.names=F)

   plot(df1, main="Newton interpolation", xlab="X", ylab="Y")
   lines(x, y)
}

inter("data.txt")

readLines("stdin", 1)

File: 4newtonInterpolation.cpp
#include<fstream>
#include<vector>
using namespace std;

void inter(string file){
    //input:
    vector<double> xin, yin;
    ifstream fin{file};
    for(double v1,v2; fin>>v1>>v2;){
        xin.push_back(v1);
        yin.push_back(v2);
    }
    fin.close();

    //finite divided differences:
    int n = xin.size();
    vector<vector<double>> M(n, vector<double>(n));
    for(int i{}; i<n; i++)
        M[i][0] = yin[i];
    for(int j{1}; j<n; j++)
        for(int i{}; i<n-j; i++)
            M[i][j] = (M[i+1][j-1] - M[i][j-1]) /
                (xin[i+j] - xin[i]);

    //interpolation:
    vector<double> x, y, factor;
    for(double i{xin[0]}; i<xin.back(); i+=0.01){
        x.push_back(i);
        y.push_back(M[0][0]);
        factor.push_back(1);
    }
    for(int j{1}; j<n; j++)
        for(int i{}; i<x.size(); i++){
            factor[i] *= x[i]-xin[j-1];
            y[i] += factor[i]*M[0][j];
        }

    //output:
    ofstream fout{"out1.txt"};
    for(int i{}; i<x.size(); i++)
        fout<<x[i]<<'\t'<<y[i]<<endl;
    fout.close();
}

int main(){
    inter("data.txt");
}
